mdtk/degradations.py

```diff
@@ -235,16 +280,35 @@ def time_shift(excerpt, min_shift=50, max_shift=np.inf, inplace=False):
 
     # Shift later
     latest_later_onset = onset + (((end_time + 1) - offset)
-                                  .clip(upper=max_shift))
+                                  .clip(upper=max_shift + 1))
     earliest_later_onset = onset + min_shift
 
+    if align_onset:
+        # Find ranges which contain a note to align to
+        # I couldn't think of a better solution than iterating here.
```

Comment:
This unfortunately adds some iteration back in a few places. To avoid this would probably require going about this a different way (first getting all possible aligned points, then checking if those are in the correct ranges). Even this, I'm not sure would work, since we have to do this calculation for each note (to shift) for each anchor point. (ie, it is O(n^2)). The current implementation does this vectorized n times (O(n), if we count vectorized as O(1)).